R/rdeephaven/src/client.cpp

```diff
@@ -0,0 +1,233 @@
+     * @param stream_ptr Pointer to an existing and populated ArrayArrayStream, populated by a call to RecordBatchReader$export_to_c(ptr) from R.
+    */
+    TableHandleWrapper* newTableFromArrowArrayStreamPtr(Rcpp::XPtr<ArrowArrayStream> stream_ptr) {
+
+        auto wrapper = internal_tbl_hdl_mngr.createFlightWrapper();
+        arrow::flight::FlightCallOptions options;
+        wrapper.addAuthHeaders(&options);
+
+        // extract RecordBatchReader from the struct pointed to by the passed tream_ptr
+        std::shared_ptr<arrow::RecordBatchReader> record_batch_reader = arrow::ImportRecordBatchReader(stream_ptr.get()).ValueOrDie();
+        auto schema = record_batch_reader.get()->schema();
+
+        // write RecordBatchReader data to table on server with DoPut
+        std::unique_ptr<arrow::flight::FlightStreamWriter> fsw;
+        std::unique_ptr<arrow::flight::FlightMetadataReader> fmr;
+        auto [new_tbl_hdl, fd] = internal_tbl_hdl_mngr.newTableHandleAndFlightDescriptor();
+        DEEPHAVEN_EXPR_MSG(wrapper.flightClient()->DoPut(options, fd, schema, &fsw, &fmr)); // TODO: need to add okOrThrow
+        while(true) {
+            std::shared_ptr<arrow::RecordBatch> this_batch;
+            DEEPHAVEN_EXPR_MSG(record_batch_reader->ReadNext(&this_batch)); // TODO: need to add ok or throw
+            if (this_batch == nullptr) {
+                break;
+            }
+            DEEPHAVEN_EXPR_MSG(fsw->WriteRecordBatch(*this_batch)); // TODO: need to add okOrThrow
+        }
+        DEEPHAVEN_EXPR_MSG(fsw->DoneWriting()); // TODO: need to add okOrThrow
+        DEEPHAVEN_EXPR_MSG(fsw->Close()); // TODO: need to add okOrThrow
+
+        return new TableHandleWrapper(new_tbl_hdl);
+    };
+
+private:
+    ClientWrapper(deephaven::client::Client ref) : internal_client(std::move(ref)) {};
+
+    const deephaven::client::Client internal_client;
+    const deephaven::client::TableHandleManager internal_tbl_hdl_mngr = internal_client.getManager();
+
+    friend ClientWrapper* newClientWrapper(const std::string &target, const ClientOptionsWrapper &client_options);
+};
+
+// factory method for calling private constructor, Rcpp does not like <const std::string &target> in constructor
+// the current implementation of passing authentication args to C++ client is terrible and needs to be redone. Only this could make Rcpp happy in a days work
+
+/**
+ * Factory method for creating a new ClientWrapper, which is responsible for maintaining a connection to the client.
+ * @param target URL that the server is running on.
+ * @param client_options A ClientOptionsWrapper containing the server connection information. See deephaven::client::ClientOptions for more information.
+ */
+ClientWrapper* newClientWrapper(const std::string &target, const ClientOptionsWrapper &client_options) {
+    return new ClientWrapper(deephaven::client::Client::connect(target, *client_options.internal_options));
+};
```

Comment:
To simplify things, in v2, we could potentially get rid of the factory methods.  They were originally introduced to control the API the user sees.  Now there is the S6 classes to do that.  Overall, this is now probably extra complexity for no gain.

Reply:
Introduced for API reasons, but kept around because Rcpp _hates_ `const std::string&` types in the ".constructor", but is happy with those types in ".factory".

Reply:
ok